### PR TITLE
feat: Add ZImage/LongCat/Sana diffusion support + LLM VL improvements…

### PR DESCRIPTION
**Pull request by er6y**
… + OpenCL fixes

Diffusion Engine:
- Add ZImageDiffusion subclass: FlowMatch Euler scheduler, PhiloxRNG noise, CLIP text encoder
- Add LongCatDiffusion subclass: LLM text encoder (lazy load), Flux-like latent packing, VAE enc/dec, T2I + Image Edit modes
- Integrate SanaDiffusion with SanaLlm into unified API, fix Euler sampling (dt/1000->dt) and CFG order
- Add DiffusionConfig/LLMEncoderConfig, GPU config (BUFFER mode, FP32, Memory_Low, OpenCL cache)
- Add DiffusionGpuMemoryMode/PrecisionMode/CFGMode enums
- Extend createDiffusion factory with full parameters for all model types
- Unify diffusion_demo to support SD1.5/Taiyi/Sana/ZImage/LongCat with cfg_scale and input_image args
- Add image processing utilities (resize, crop, colorspace, pack/unpack latents)
- Implement FlowMatchEuler scheduler

LLM / Vision:
- omni.cpp: Qwen3-VL vision fixes (floor rounding, nullptr check)
- omni.hpp: mrope position ids fix max(T,H,W)+1
- tokenizer.hpp: public wrapper header, MNN_PUBLIC export
- llm_demo: LLM_DEMO_ONELINE mode
- pymnn llm.h: forward_all() binding
- Fix Qwen2_5Vision transformer_fuse for window attention compatibility

OpenCL fixes:
- BinaryBufExecution: localWorkSize divisibility check
- binary_buf.cl: float4/int4 init, per-element ReLU (NVIDIA compiler bug workaround)

Other fixes:
- ShapeSliceTf/ShapeWhere: shape calculation rewrite
- OnnxEinsum: outer product broadcast _Unsqueeze fix
- Pipeline: NaN/Inf debug check macro (disabled by default)
- Fix LongCat unpackLatentsGPU to match master implementation

### CODE REVIEW

File: source/shape/ShapeSliceTf.cpp [hunk 1]
```diff
@@ -1,62 +1,67 @@
-//
-//  ShapeSliceTf.cpp
```
Comment (wangzhaode) [inline source/shape/ShapeSliceTf.cpp:2]: 这个文件如果没有修改请从PR里移除

File: source/shape/ShapeWhere.cpp [hunk 2]
```diff
@@ -1,71 +1,73 @@
-//
```
Comment (wangzhaode) [inline source/shape/ShapeWhere.cpp:1]: 这个文件如果没有修改请从PR里移除

File: source/backend/opencl/core/BufferPool.cpp [hunk 3]
```diff
@@ -51,13 +60,20 @@ void BufferPool::clear() {
 }
 
 void BufferPool::releaseFreeList() {
+    std::multimap<size_t, std::shared_ptr<OpenCLBufferNode>> keepList;
```
Comment (Qxinyu) [inline source/backend/opencl/core/BufferPool.cpp]: Using this logic in an OpenCL dynamic memory pool might result in the release of memory reused by preceding layers, leading to a crash.
Do you mean to release the memory previously allocated in the DYNAMIC_IN_EXECUTION memory pool used by the Attention operator?

**Comment by er6y** [comment]
Thanks for the review! To clarify:

my modification only affects [BufferPool::releaseFreeList()]
BufferPool.cpp:61:0-76:1) (static pool), NOT [BufferExecutionPool::releaseFreeList()](BufferPool.cpp:61:0-76:1) (DYNAMIC_IN_EXECUTION pool).

What i changed:
BufferPool (Line 62-77): Keep buffers >1GB in the free list after [releaseFreeList()](BufferPool.cpp:61:0-76:1) is called
BufferExecutionPool(Line 135-143): No changes

The >1GB threshold targets large attention score tensors (6.78GB in our case). These are temporary intermediate tensors that are allocated and freed within a single inference pass, not persistent buffers shared across layers.

The actual fix for layer reuse is in Pipeline.cpp:
The critical fix is in Pipeline.cpp (Line 1012-1049), where we force-release large tensors with useCount<=1 before allocating new large outputs. This ensures the previous block's attention score buffer is recycled before the next one is allocated, preventing two 6.78GB tensors from coexisting.

The BufferPool.cpp change is just an optimization to avoid repeated alloc/free of 6.78GB buffers across multiple diffusion blocks.

Does this makes sense or not?

**Comment by Qxinyu** [comment]
> Thanks for the review! To clarify:
> 
> my modification only affects [BufferPool::releaseFreeList()] BufferPool.cpp:61:0-76:1) (static pool), NOT BufferExecutionPool::releaseFreeList() (DYNAMIC_IN_EXECUTION pool).
> 
> What i changed: BufferPool (Line 62-77): Keep buffers >1GB in the free list after releaseFreeList() is called BufferExecutionPool(Line 135-143): No changes
> 
> The >1GB threshold targets large attention score tensors (6.78GB in our case). These are temporary intermediate tensors that are allocated and freed within a single inference pass, not persistent buffers shared across layers.
> 
> The actual fix for layer reuse is in Pipeline.cpp: The critical fix is in Pipeline.cpp (Line 1012-1049), where we force-release large tensors with useCount<=1 before allocating new large outputs. This ensures the previous block's attention score buffer is recycled before the next one is allocated, preventing two 6.78GB tensors from coexisting.
> 
> The BufferPool.cpp change is just an optimization to avoid repeated alloc/free of 6.78GB buffers across multiple diffusion blocks.
> 
> Does this makes sense or not?

Thanks for the detailed explanation, I understand the situation now.

Just to confirm two points:

This modification indeed does not affect the dynamic memory pool, correct?
Are the large attention score tensors you mentioned (6.78GB) actually allocated from the static memory pool?

**Comment by er6y** [comment]
> > Thanks for the review! To clarify:
> > my modification only affects [BufferPool::releaseFreeList()] BufferPool.cpp:61:0-76:1) (static pool), NOT BufferExecutionPool::releaseFreeList() (DYNAMIC_IN_EXECUTION pool).
> > What i changed: BufferPool (Line 62-77): Keep buffers >1GB in the free list after releaseFreeList() is called BufferExecutionPool(Line 135-143): No changes
> > The >1GB threshold targets large attention score tensors (6.78GB in our case). These are temporary intermediate tensors that are allocated and freed within a single inference pass, not persistent buffers shared across layers.
> > The actual fix for layer reuse is in Pipeline.cpp: The critical fix is in Pipeline.cpp (Line 1012-1049), where we force-release large tensors with useCount<=1 before allocating new large outputs. This ensures the previous block's attention score buffer is recycled before the next one is allocated, preventing two 6.78GB tensors from coexisting.
> > The BufferPool.cpp change is just an optimization to avoid repeated alloc/free of 6.78GB buffers across multiple diffusion blocks.
> > Does this makes sense or not?
> 
> Thanks for the detailed explanation, I understand the situation now.
> 
> Just to confirm two points:
> 
> This modification indeed does not affect the dynamic memory pool, correct? Are the large attention score tensors you mentioned (6.78GB) actually allocated from the static memory pool?



Yes, both points are correct:

1. Static pool only
My modification only affects the static pool (BufferPool::releaseFreeList()), not the dynamic pool (BufferExecutionPool).

Static pool (BufferPool.cpp:62-77): Added >1GB threshold to keep large buffers in free list
Dynamic pool (BufferExecutionPool.cpp:135-143): No changes at all
2. Large tensors are allocated from static pool
The 6.78GB attention score tensors are indeed allocated from the static pool, not the dynamic pool.

Why? These tensors are:

Temporary intermediate results (attention scores in diffusion blocks)
Allocated/freed within a single inference pass (not persistent across layers)
Managed by BufferPool (static pool), which handles short-lived allocations
The dynamic pool (BufferExecutionPool) is only used for persistent layer outputs that need to be shared across multiple pipeline stages.

Key point
The real fix is in Pipeline.cpp (Line 1012-1049), which forces recycling of large tensors before allocating new ones. The BufferPool.cpp change is just an optimization to avoid repeated alloc/free overhead for these 6.78GB buffers.

File: transformers/llm/engine/src/llm.cpp [hunk 4]
```diff
@@ -491,6 +491,22 @@ std::vector<Express::VARP> Llm::forwardRaw(Express::VARP hiddenState, Express::V
         mContext->status = LlmStatus::INTERNAL_ERROR;
         return outputs;
     }
+    // Save hidden_states to file for LongCat text encoder
```
Comment (wangzhaode) [inline transformers/llm/engine/src/llm.cpp]: remove th debug code

File: transformers/llm/engine/src/omni.cpp [hunk 5]
```diff
@@ -175,9 +175,15 @@ std::vector<int> Omni::qwen2VisionProcess(VARP image) {
     constexpr int temporal_patch_size = 2;
     constexpr int merge_size = 2;
     const int align_size = patch_size * merge_size;
+    MNN_PRINT("[Omni] qwen2VisionProcess: patch_size=%d, align_size=%d, hasWindowIndex=%d, isQwen3VL=%d\n",
```
Comment (wangzhaode) [inline transformers/llm/engine/src/omni.cpp]: remove DEBUG INFO

File: transformers/llm/engine/src/omni.cpp [hunk 6]
```diff
@@ -340,8 +347,14 @@ std::vector<int> Omni::qwen2VisionProcess(VARP image) {
     attention_mask->setName("attention_mask");
     MNN::Express::Variable::save({patches, position_ids, attention_mask}, "input.mnn");
 #endif
+    MNN_PRINT("[Omni] qwen2VisionProcess: running mVisionModule->onForward with %d inputs\n", (int)moduleInputs.size());
     auto outputs = mVisionModule->onForward(moduleInputs);
+    MNN_PRINT("[Omni] qwen2VisionProcess: mVisionModule returned %d outputs\n", (int)outputs.size());
     auto imageEmbedding = outputs[0];
+    auto embInfo = imageEmbedding->getInfo();
```
Comment (wangzhaode) [inline transformers/llm/engine/src/omni.cpp]: REMOVE DEBUG INFO

File: source/shape/ShapeSliceTf.cpp (hunk 1 above)
Comment (wangzhaode) [reply source/shape/ShapeSliceTf.cpp:2]: 这个文件还没有移除

**Comment by wangzhaode** [comment]
@er6y Please resolve the conflicts and comments.

**Comment by er6y** [comment]
> @er6y Please resolve the conflicts and comments.
Resolved +2...

**Review by wangzhaode — CHANGES_REQUESTED** [review]
Hi @er6y, several review comments need to be addressed before this PR can move forward. Please see the inline comments for details:

1. Remove unchanged files (ShapeSliceTf.cpp, ShapeWhere.cpp)
2. Remove debug code in llm.cpp and omni.cpp
3. Address the OpenCL BufferPool concern raised by @Qxinyu

Also, given the large scope of this PR (50 files across Diffusion/LLM/OpenCL/Converter), it would be much easier to review and merge if split into smaller, focused PRs. Thanks!

File: source/shape/ShapeSliceTf.cpp (hunk 1 above)
Comment (er6y) [reply source/shape/ShapeSliceTf.cpp:2]: ShapeSliceTf.cpp：新增了GPU backend兼容逻辑——当 host<int32_t>() 返回 nullptr（OpenCL/Vulkan后端无法直接读host内存）时，用保守估计（full input size）替代精确计算。建议保留？

File: source/shape/ShapeWhere.cpp (hunk 2 above)
Comment (er6y) [reply source/shape/ShapeWhere.cpp:1]: ShapeWhere.cpp：同样，GPU backend下 host<void>() 为 nullptr 时，原代码会把输出dim设为0（导致模型运行错误），修改后保留保守估计 elementSize。建议保留？

File: transformers/llm/engine/src/llm.cpp (hunk 4 above)
Comment (er6y) [reply transformers/llm/engine/src/llm.cpp]: 接受移除。后续移除，再rebase

File: transformers/llm/engine/src/omni.cpp (hunk 6 above)
Comment (er6y) [reply transformers/llm/engine/src/omni.cpp]: 接受移除

**Comment by er6y** [comment]
> Hi @er6y, several review comments need to be addressed before this PR can move forward. Please see the inline comments for details:
> 
> 1. Remove unchanged files (ShapeSliceTf.cpp, ShapeWhere.cpp)
> 2. Remove debug code in llm.cpp and omni.cpp
> 3. Address the OpenCL BufferPool concern raised by @Qxinyu
> 
> Also, given the large scope of this PR (50 files across Diffusion/LLM/OpenCL/Converter), it would be much easier to review and merge if split into smaller, focused PRs. Thanks!  
---> got it, many thanks for the reviews and comments!!

关于问题3：
3. Address the OpenCL BufferPool concern raised by @Qxinyu
alloc失败时 releaseFreeList() 后重试一次
releaseFreeList() 保留 >1GB 的大buffer不释放（"用于1024 edit mode的6.78GB fp32 attention score tensors"）
这个修改有实际场景支撑（diffusion大分辨率编辑需要超大buffer），但硬编码1GB阈值，这个地方确实屎，可能是导致concern 点。是否有更好建议？

我建议改进如下，我会变更 两层重试 + 动态阈值 方案 如果测试通过会重新刷新PR，也请review如下是否ok：
- releaseFreeList() 恢复 全清语义 —— onGabageCollect() 调用时正常回收所有 buffer，不再有内存泄漏隐患。
- alloc() 分配失败时改为 两层重试 ：
  - 第一层（部分释放） ：只释放小于 max(请求大小, kMinKeepThreshold) 的 buffer。大 buffer（如 diffusion 多步去噪中复用的 6.78GB attention score tensor）会被保留，避免反复分配/释放的开销。 kMinKeepThreshold 为命名常量（256MB），附有注释说明。
  - 第二层（全部释放） ：如果第一层释放后仍不够，则彻底释放所有 buffer（包括大 buffer），确保不会 OOM 卡死。
- 此修改仅影响 静态池 （ BufferPool ），不影响动态池（ BufferExecutionPool ，即 DYNAMIC_IN_EXECUTION 使用的池）。

File: transformers/llm/engine/src/omni.cpp (hunk 5 above)
Comment (er6y) [reply transformers/llm/engine/src/omni.cpp]: 接受移除

**Comment by wangzhaode** [comment]
感谢你持续完善这个 PR，也感谢你对之前评审意见的认真回复。

我们重新检查了最新版本。目前这个 PR 在一个提交中同时包含 Diffusion、LLM、OpenCL、Converter、Shape 和 PyMNN 等多个模块，共涉及 50 个文件，整体范围仍然过大，很难完成可靠的代码评审和回归验证。另外，当前 Pipeline 中提前释放大 Tensor 的逻辑存在正确性风险，OpenCL 内存池修改也还需要独立的压力测试；现有检查只有 CLA，没有覆盖实际构建和相关运行路径。

为了让有价值的修改能够更快合入，建议基于最新 master 将内容拆成若干独立 PR，例如每个 Diffusion 模型分别提交，OpenCL 内存优化单独提交并附测试结果，Converter、Shape 和 PyMNN 修改也分别提交。我们会更容易逐项评审和合并。再次感谢你的投入和理解！
